x-pack/plugins/lens/public/lens_suggestions_api/helpers.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { VisualizeFieldContext } from '@kbn/ui-actions-plugin/public';
+import type { VisualizeEditorContext, Suggestion } from '../types';
+import type { TypedLensByValueInput } from '../embeddable/embeddable_component';
+
+/**
+ * Returns the suggestion updated with external visualization state for ES|QL charts
+ * The visualization state is merged with the suggestion if the datasource is textBased, the columns match the context and the visualization type matches
+ * @param suggestion the suggestion to be updated
+ * @param visAttributes the preferred visualization attributes
+ * @param context the lens suggestions api context as being set by the consumers
+ * @returns updated suggestion
+ */
+
+export function mergeSuggestionWithVisContext({
+  suggestion,
+  visAttributes,
+  context,
+}: {
+  suggestion: Suggestion;
+  visAttributes: TypedLensByValueInput['attributes'];
+  context: VisualizeFieldContext | VisualizeEditorContext;
+}): Suggestion {
+  if (
+    visAttributes.visualizationType !== suggestion.visualizationId ||
+    !('textBasedColumns' in context)
+  ) {
+    return suggestion;
+  }
+
+  // it should be one of 'formBased'/'textBased' and have value
+  const datasourceId: 'formBased' | 'textBased' | undefined = [
+    'formBased' as const,
+    'textBased' as const,
+  ].find((key) => Boolean(visAttributes.state.datasourceStates[key]));
```

Comment:
nit: Is the third time you are using this function, is probably useful to add it an utility function

Reply:
Sure I added one on the vis utils https://github.com/elastic/kibana/pull/197453/commits/1e92f8de1c0b67da606261853c0c3ceccb4351d8